src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,320 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "pal_types.h"
+#include "pal_utilities.h"
+#include "pal_gssapi.h"
+
+#if HAVE_GSSFW_HEADERS
+#include <GSS/GSS.h>
+#else
+#include <gssapi/gssapi_ext.h>
+#include <gssapi/gssapi_krb5.h>
+#endif
+
+#include <assert.h>
+#include <string.h>
+
+static_assert(PAL_GSS_C_DELEG_FLAG == GSS_C_DELEG_FLAG, "");
+static_assert(PAL_GSS_C_MUTUAL_FLAG == GSS_C_MUTUAL_FLAG, "");
+static_assert(PAL_GSS_C_REPLAY_FLAG == GSS_C_REPLAY_FLAG, "");
+static_assert(PAL_GSS_C_SEQUENCE_FLAG == GSS_C_SEQUENCE_FLAG, "");
+static_assert(PAL_GSS_C_CONF_FLAG == GSS_C_CONF_FLAG, "");
+static_assert(PAL_GSS_C_INTEG_FLAG == GSS_C_INTEG_FLAG, "");
+static_assert(PAL_GSS_C_ANON_FLAG == GSS_C_ANON_FLAG, "");
+static_assert(PAL_GSS_C_PROT_READY_FLAG == GSS_C_PROT_READY_FLAG, "");
+static_assert(PAL_GSS_C_TRANS_FLAG == GSS_C_TRANS_FLAG, "");
+static_assert(PAL_GSS_C_DCE_STYLE == GSS_C_DCE_STYLE, "");
+static_assert(PAL_GSS_C_IDENTIFY_FLAG == GSS_C_IDENTIFY_FLAG, "");
+static_assert(PAL_GSS_C_EXTENDED_ERROR_FLAG == GSS_C_EXTENDED_ERROR_FLAG, "");
+static_assert(PAL_GSS_C_DELEG_POLICY_FLAG == GSS_C_DELEG_POLICY_FLAG, "");
+
+static_assert(PAL_GSS_COMPLETE == GSS_S_COMPLETE, "");
+static_assert(PAL_GSS_CONTINUE_NEEDED == GSS_S_CONTINUE_NEEDED, "");
+
+#if !HAVE_GSS_SPNEGO_MECHANISM
+static char gss_spnego_oid_value[] = "\x2b\x06\x01\x05\x05\x02"; // Binary representation of SPNEGO Oid (RFC 4178)
+#endif
+
+// transfers ownership of the underlying data from gssBuffer to PAL_GssBuffer
+static void NetSecurityNative_MoveBuffer(gss_buffer_t gssBuffer, struct PAL_GssBuffer* targetBuffer)
+{
+    assert(gssBuffer != nullptr);
+    assert(targetBuffer != nullptr);
+
+    targetBuffer->length = gssBuffer->length;
+    targetBuffer->data = static_cast<uint8_t*>(gssBuffer->value);
+}
+
+static uint32_t NetSecurityNative_AcquireCredSpNego(uint32_t* minorStatus,
+                                                    GssName* desiredName,
+                                                    gss_cred_usage_t credUsage,
+                                                    GssCredId** outputCredHandle)
+{
+    assert(minorStatus != nullptr);
+    assert(desiredName != nullptr);
+    assert(outputCredHandle != nullptr);
+
+#if HAVE_GSS_SPNEGO_MECHANISM
+    gss_OID_set_desc gss_mech_spnego_OID_set_desc = {.count = 1, .elements = GSS_SPNEGO_MECHANISM};
+#else
+    gss_OID_desc gss_mech_spnego_OID_desc = {.length = 6, .elements = static_cast<void*>(gss_spnego_oid_value)};
```

Comment:
Is there a way to make the compiler produce this 6 safely, rather than trusting a human? `sizeof(gss_spnego_oid_value)`, maybe?  (Though that might be 7? I'm not sure if it uses the trailing null... but then the assignment could be rewritten to use { ... } instead of " ... ")